pkg/wasm/client.go

```diff
@@ -0,0 +1,260 @@
+	}
+	img, err := mutate.AppendLayers(empty.Image, layer)
+	if err != nil {
+		return fmt.Errorf("failed to append image layers: %w", err)
+	}
+	img = mutate.MediaType(img, types.DockerManifestSchema2)
+	img, err = mutate.CreatedAt(img, v1.Time{Time: time.Now()})
+	if err != nil {
+		return fmt.Errorf("failed to set created time for the image: %w", err)
+	}
+	hash, err := img.Digest()
+	if err != nil {
+		return fmt.Errorf("failed to get digest of the image: %w", err)
+	}
+
+	// Push the image to remote
+	if err = remote.Write(tag.Repository.Digest(hash.String()), img, options...); err != nil {
+		return fmt.Errorf("failed to push image %s: %w", tag, err)
+	}
+	fmt.Printf("digest of the image: %v\n", hash.String())
+
+	// Construct the image index.
+	var index v1.ImageIndex = empty.Index
+	if rmt != nil {
+		index, err = rmt.ImageIndex()
+		if err != nil {
+			return fmt.Errorf("unable to parse an image as image index: %w", err)
+		}
+	}
+	indexManifest, err := index.IndexManifest()
+	if err != nil {
+		return fmt.Errorf("unable to parse image manifest: %w", err)
+	}
+	foundWasmPlatform := false
+	wasmPlatform := v1.Platform{
+		Architecture: "wasm",
+		OS:           "js",
+	}
+
+	// Check if there is a manifest for wasm/js platform. If found, we can simply
+	// update its digest to the one we just pushed. Otherwise, we append a new
+	// manifest to the manifest list.
+	for i, manifest := range indexManifest.Manifests {
+		if manifest.Platform != nil && manifest.Platform.Architecture == wasmPlatform.Architecture && manifest.Platform.OS == wasmPlatform.OS {
+			foundWasmPlatform = true
+			indexManifest.Manifests[i].Digest = hash
+		}
+	}
+	if !foundWasmPlatform {
+		index = mutate.IndexMediaType(mutate.AppendManifests(index, mutate.IndexAddendum{
+			Add: img,
+			Descriptor: v1.Descriptor{
+				MediaType: "application/vnd.docker.distribution.manifest.v2+json",
+				Digest:    hash,
+				Platform:  &wasmPlatform,
+			},
+		}), types.DockerManifestList)
+	}
+
+	// Push the image index to remote.
+	err = remote.WriteIndex(tag, index, options...)
+	if err != nil {
+		return fmt.Errorf("unable to write image index: %w", err)
+	}
+	indexHash, err := index.Digest()
+	if err != nil {
+		return fmt.Errorf("unable to get digest for the image index: %w", err)
+	}
+	fmt.Printf("digest of the image index: %v\n", indexHash.String())
+	return nil
+}
+
+func wasmFileToTar(filename string) (io.Reader, error) {
+	buf := bytes.NewBuffer(nil)
+	writer := tar.NewWriter(buf)
+
+	b, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read from file %v: %w", filename, err)
+	}
+	blen := len(b)
+
+	if err = writer.WriteHeader(&tar.Header{
+		Name: filepath.Base(filename),
+		Size: int64(blen),
+		Mode: 0644,
+	}); err != nil {
+		return nil, fmt.Errorf("failed to write tar header: %w", err)
+	}
+
+	if n, err := writer.Write(b); err != nil {
+		return nil, fmt.Errorf("failed to write tar contents: %w", err)
+	} else if n != blen {
```

Comment:
I don't think you need to check this, I think the io.Writer contract says that err != nil if it's a short-write (?)

Reply:
You are right! Removed.